### PR TITLE
GH-61: Add link and list package

### LINKED ISSUE #59: List

### LINKED ISSUE #61: Links
This is tracked by #57 but some discussion could go here. I am thinking the syntax will be an optional label argument and if one is provided it is used otherwise the link is used as the label. Only thing that could be messy is quotes or other characters that mess with the output language in the URL

### PR DESCRIPTION
**Pull request by axoxelol1**
Resolves #59 and #61 by adding a list and link package. There is only HTML support as of now, so if this gets merged issues for adding LaTeX support should be created. Probably should have split this up into two PRs but the link package was so small I just kept going, if it is a big deal I can split it up now. Could have made the commits reference the larger issues (#57) instead maybe, not sure how to work with small issues in the best way.

Please test the packages to make sure they are working, I have been sick and tired when finishing up :p

- Resolves #59
- Resolves #61

### CODE REVIEW

**Review by Forsinge — CHANGES_REQUESTED** [review]
Nice work, especially with the nested lists!

I found a few things to comment on, nothing too important so feel free to just mark them as resolved if you don't feel they are worth fixing.

Tests are also quite nice to have, though I see our LaTeX package do not have any so maybe there's no reason to enforce having them.

**Comment by CMDJojo** [comment]
Additionally, it seems like it doesn't work to number list items 1., 2. and 3., just to number them 1., 1. and 1.. See screenshots. To me, it seems that it should work. You may look at Djot, they allow any numbers, so you may number a list `4. 3. 11.` to get `1. 2. 3.`, or enforce a strict order, but `1. 2. 3.` should at least be okay in my mind.
Screenshot 2023-03-08 at 16 35 31

**Comment by CMDJojo** [comment]
This also resolves #61, but the PR body wasn't formatted so GitHub picked up on it. Edited your PR body to work correctly

File: packages/list/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        let items: Vec<ListItem> =
+            s.lines()
+                .skip_while(|&l| l == "")
+                .fold(Vec::new(), |mut items, line| {
+                    let start = line
+                        .trim()
+                        .split_ascii_whitespace()
+                        .next()
+                        .unwrap_or_else(|| "")
+                        .parse::<ItemType>();
```
[the remark this answers is not in the document]
Comment (axoxelol1) [reply packages/list/src/lib.rs:118]: Might have a look at this but feels like a lower priority, fixing other stuff and clippy comments right now =)

**Comment by CMDJojo** [comment]
Just a heads-up: it seems like you have gotten a merge commit into your branch. This is not good, since merge commits are not allowed in `main`. I know that I have gotten a lot of problems when I accidentally got one in my branch. Try to rebase again to flatten your branch.

**Comment by axoxelol1** [comment]
Ye no worries I am aware, fixed a cargo.toml conflict in the web editor and it merged, wanted to see if it would rebase. I'll fix it =).

On another note djot seems to ignore all but the first list item, and it starts the order from that. So 4. 1. 2. becomes 4. 5. 6. Seems reasonable so I will try and implement that.

**Comment by CMDJojo** [comment]
> Ye no worries I am aware, fixed a cargo.toml conflict in the web editor and it merged, wanted to see if it would rebase. I'll fix it =).
> 
> On another note djot seems to ignore all but the first list item, and it starts the order from that. So 4. 1. 2. becomes 4. 5. 6. Seems reasonable so I will try and implement that.

Yeah I noticed that later when I tried it out. I think it is to allow "broken" lists like:
```
1. This is one item
2. This is a second item
and here is some text explaining the second item
3. Here is a third item, numbered 3
```

**Review by Forsinge — APPROVED** [review]
Nice changes, solid work!

**Comment by CMDJojo** [comment]
Really good job! I tested it a bit in preview, I think it looks very good! The only two things I would comment on is:
* It would be nice not having to have a space (I think the dot as a separator in `1.`, `2.` is verbose enough so we don't really need a space as well). I know you have talked about this earlier, and if you think it is too much of a hassle to fix, feel free to skip it!
* When actually using it, I think four spaces are too much to need as an indent. I think two spaces is enough and that it is most likely more annoying than helpful to need to have four spaces.

But as said, good work! I love how you can change numbering mid-list by doing
```
I. This is numbered "I"
I. This is numbered "II"
1. This is numbered "3"
```
I don't know if it is intentional or not but it is a fun easter-egg. This, unfortanely, breaks down if you do this:
```
a. This is labeled "a"
b. This is labeled "b"
c. This is labeled "iii" since it treats the "c" as a roman numeral
```
I don't know if that is too hard to fix or not, and what the best way to fix it is. If you don't think it is worth the effort, skip it. But good job with the module!

**Comment by Forsinge** [comment]
I'm not sure I agree with your second point, since you get 4 or more spaces using tab in many if not most editors. In our playground, tab works just fine with this version of list. 

I suppose you could add an indent argument so that the user can customize it to their needs.

**Comment by CMDJojo** [comment]
I am unsure myself, and yes I know most IDEs default to 4 spaces. My initial thought was that that it took some time to write 4 spaces in the playground, didn’t notice you could use tabs, so that was nice! Keep it at 4 then :)

**Comment by CMDJojo** [comment]
What about the numbering "issue" when doing a. b. c.? Is that something fixable/worth fixing before merging/something intentional? We could just disallow such label switching, or just allow it after a broken list, not mid-list, maybe? Or if it is too troublesome, skip fixing it!

**Comment by adelhult** [comment]
I can chip in and say great work as well! :smile: 

I have a question about this though. It was a long time since I dealt with lists in HTML, but this should only be a `<ul>` and a `<ol>` right? It seems like a new table is created for each table item  + a extra at the start.
image

**Comment by axoxelol1** [comment]
Hm, I think I will skip the . or ) delimiter for now and use the space, does not feel like it is worth spending time on it when there are other issues that can be handled. I used 4 spaces due to tab but I agree that adding an indent argument would be nice, I will do that today it _should_ be easy.

I am not sure how we should handle i) or c) or m).... These are both valid alpha and roman so there could be cases where you want one or the other. The easiest would be to remove lower roman numerals and only allow upper but that seems like a dumb solution. I like the fact that you can switch in the list, djot does not allow this and that way avoids it. The only time this is an issue if you want to start a list with c or i as an alpha. Especially for c this might be something you want to do. Could add an argument to say what should take precedence.

@adelhult ye it would give the same result with one <ol> tag and multiple <li>. The reason it keeps creating lists is because this was the easiest way to make sure that the starting and types are right, but if the type is the same it would be enough with just another <li>. Does not seem to be a problem even though it is a bit unconventional.

**Comment by CMDJojo** [comment]
If we disallow switching numbering system mid-list, it would not confuse ”a b c”, or if we check if it matches the ”normal” next-following tag (if we have ”a. b.” we check if next is ”c.” and then never change the numbering) would probably work as well. Don’t know what is best

**Comment by axoxelol1** [comment]
I think a. b. c. is fine because you can do a. a. a. for the same result, the only problem is starting a list with c or i

**Comment by axoxelol1** [comment]
Added indent argument but found a bug if the list starts with a bullet then uses a number it will not start with that. Will fix tomorrow.

**Comment by CMDJojo** [comment]
Great work! I don't know if the indent argument is necessary, but it can't be bad. The only thing I feel might be problematic is that, as stated before,
```
[list]
a. Item
b. Item
c. Item
d. Item
e. Item
```
gives this:
```
a. Item
b. Item
iii. Item
iv. Item
e. Item
```
which is unexpected. But I see your argument above that `a. b. c. d. e.` is still expressible by `a. a. a. a. a.`, and if you feel like it is not worth fixing, don't. Otherwise, amazing work!

**Comment by axoxelol1** [comment]
It's not that I don't think its worth fixing, I'm just not sure how I would go about it. I like having the ability to switch types in the list. I would like to be able to start a list with c or d though and I think that will be much more common than wanting  to start with the roman numeral c so I think I might just make alpha take precedence over roman

**Pull request merged**